[RFC] Job control: enable mouse after foregrounding

mhinz opened the pull request:
After foregrounding a neovim process, the terminal emulator's mouse mode gets
enabled according to the 'mouse' option.

References #2335.

mhinz committed:
Job control: reset mouse support after foregrounding

Reset mouse support to whatever was used before the process was suspended.

References #2335.

Pyrohh commented:
Assuming I'm understanding the new UI architecture (it seems really easy to follow), then LGTM.

tarruda commented:
While this is fine for the current version, I'd like to decouple the TUI from all state kept in the core, which `setmouse` depends on.

@mhinz could you change this PR so that the mouse enable/disable state is kept into the `TUIData` structure? This will be necessary to make it work with remote nvim instances.

mhinz commented:
@tarruda I split up `setmouse()` into two functions: `setmouse()` and `can_enable_mouse()`. And I added a member `bool can_use_mouse` to `TUIData`. That member is initialized in `tui_start()` via `can_enable_mouse()`.

I hope this is the change you had in mind, but I'm lacking a general overview, so it might be the wrong approach. :-)

tarruda commented:
@mhinz What I meant is that the process running tui.c can be different than the process running mouse.c, so it shouldn't call core functions like you did. A way fix it would be to expose `can_enable_mouse` via msgpack-rpc, but that would be an overkill for something that is only used by the UI.

The simplest solution is to store the  `tui_mouse_{on,off}` state and restore it from `tui_suspend`, something like this:

```c
static void tui_suspend(UI *ui)
{
  TUIData *data = ui->data;
  bool should_enable_mouse = data->mouse_enabled;
  tui_stop(ui);
  kill(0, SIGTSTP);
  UI* ui = tui_start();
  if (should_enable_mouse) {
    tui_mouse_on(ui);
  } 
}
```

Where `mouse_enabled` is set by `tui_mouse_on`/`tui_mouse_off`. This needs to change `tui_start` to return the new UI instance.

mhinz commented:
@tarruda Oh, I see. It should work as expected now.

splinterofchaos commented on src/nvim/tui/tui.c:
```diff
@@ -561,9 +564,14 @@ static void tui_flush(UI *ui)
+  ui = tui_start();
+  if (enable_mouse) {
+    tui_mouse_on(ui);
+  };
```
superfluous semicolon

mhinz replied:
*cough*

splinterofchaos commented:
LGTM :+1:
